server/src/main/java/org/cloudfoundry/identity/uaa/zone/VmidentityIdentityZoneProvisioning.java

```diff
@@ -79,42 +79,84 @@ public IdentityZone retrieveBySubdomain(String subdomain) {
     @Override
     public List<IdentityZone> retrieveAll() {
         try {
-            String systemTenant = this._idmClient.getSystemTenant();
             ArrayList<IdentityZone> list = new ArrayList<IdentityZone>();
             for (String tenant : this._idmClient.getAllTenants()) {
-                list.add(identityZoneForTenant(this._idmClient.getTenant(tenant),
-                        systemTenant.equalsIgnoreCase(tenant)));
+                list.add(getIdentityZoneFromTenant(tenant));
             }
-
+            
             return list;
         } catch (Exception ex) {
             throw new IllegalStateException("Failed to retrieve zones.", ex);
         }
     }
+    
+    private IdentityZone identityZoneForTenant(String tenantName) throws Exception {
+        IdentityZone zone = new IdentityZone();
+        Tenant t = this._idmClient.getTenant(tenantName);
+        
+        zone.setId(t.getName());
+        zone.setName(t.getName());
+        zone.setSubdomain(t.getName());
+        zone.setDescription(t._longName);
+        zone.setVersion(1);
+        
+        return zone;
+    }
+    
+    private void configureIdentityZone(String tenantName, IdentityZone zone) throws Exception {
+        IdentityZoneConfiguration config = this._configs.get(tenantName);
+        if (config == null) {
+            config = new IdentityZoneConfiguration();
+            this._configs.put(tenantName, config);
+        }
+        
+        Map<String, String> keys = new HashMap<String, String>();
+        PrivateKey privateKey = _idmClient.getTenantPrivateKey(tenantName);
+        StringWriter writer = new StringWriter();
+        PEMWriter pemWriter = new PEMWriter(writer);
+        pemWriter.writeObject(privateKey);
+        pemWriter.close();
+        
+        keys.put("key-id-1", writer.toString());
+        config.getTokenPolicy().setKeys(keys);
+        config.getTokenPolicy().setActiveKeyId("key-id-1");
+        config.getTokenPolicy().setAccessTokenValidity(
+                Ints.saturatedCast(_idmClient.getMaximumBearerTokenLifetime(tenantName)));
+        config.getTokenPolicy().setRefreshTokenValidity(
+                Ints.saturatedCast(_idmClient.getMaximumBearerRefreshTokenLifetime(tenantName)));
```

Comment:
There is - I have the change for that sitting here, but haven't finished it just yet. It seems to also want a private key password (`SamlConfig#setPrivateKeyPassword(String)`) and possibly a separate certificate. I'm trying to suss out whether we need the password (I don't think we do). And what the certificate is for.